grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -51,14 +149,615 @@ class ApiGetOpenApiDescriptionHandler(api_call_handler_base.ApiCallHandler):
+          issubclass(args_type, rdf_structs.RDFProtoStruct)
+      ):
+        raise TypeError("Router method args type is not a RDFProtoStruct "
+                        "subclass.")
+      field_descriptors = args_type.protobuf.DESCRIPTOR.fields
+
+    # Separate fields into params: path, query and part of the request body.
+    path_params_names = set(self._GetPathArgsFromPath(path))
+    path_params = []
+    query_params = []
+    body_params = []
+    for field_d in field_descriptors:
+      if field_d.name in path_params_names:
+        path_params.append(field_d)
+      elif http_method.upper() in ("GET", "HEAD"):
+        query_params.append(field_d)
+      else:
+        body_params.append(field_d)
+
+    return path_params, query_params, body_params
+
+  def _GetOperationDescription(
+      self,
+      http_method: str,
+      path: str,
+      router_method: Any,
+  ) -> Dict[str, Any]:
+    """Create the OpenAPI Operation Object associated with the given args."""
+
+    url_path = (  # Rewrite the path in a URL-friendly format.
+      path
+        .replace('/', '-')
+        .replace('<', '_')
+        .replace('>', '_')
+        .replace(':', '-')
+    )
+    path_params, query_params, body_params = (
+      self._SeparateFieldsIntoParams(http_method, path, router_method.args_type)
+    )
+
+    # The Operation Object associated with the current http method.
+    operation_obj = {
+      "tags": [router_method.category or "NoCategory", ],
+      "description": router_method.doc or "No description.",
+      "operationId": urlparse.quote(f"{http_method}-{url_path}-"
+                                    f"{router_method.name}"),
+      "parameters": self._GetParameters(path_params, query_params),
+      "responses": {
+        "200": (
+          self._GetResponseObject200(router_method.result_type,
+                                     router_method.name)
+        ),
+        "default": self._GetResponseObjectDefault(router_method.name),
+      },
+    }
+    if body_params: # Only POST methods should have an associated "requestBody".
+      operation_obj["requestBody"] = self._GetRequestBody(body_params)
+
+    return operation_obj
+
+  def _GetPaths(self) -> Dict[str, Dict]:
+    """Create the OpenAPI description of all the routes exposed by the API."""
+
+    # The Paths Object "paths" field of the root OpenAPI object.
+    paths_obj = dict()  # type: Dict[str, Dict]
+
+    router_methods = self.router.__class__.GetAnnotatedMethods()
+    for router_method_name in router_methods:
+      router_method = router_methods[router_method_name]
+      for http_method, path, strip_root_types in router_method.http_methods:
+        normalized_path = self._NormalizePath(path)
+        path_args = set(self._GetPathArgsFromPath(path))
+
+        if normalized_path not in paths_obj:
+          paths_obj[normalized_path] = dict()
+
+        # The Path Object associated with the current path.
+        path_obj = paths_obj[normalized_path]
```

Comment:
`path_obj = paths_obj.setdefault(normalized_path, dict())`, or even better: just use `paths_obj = collections.defaultdict(dict)`.

Reply:
Ooh, that's awesome! Thank you for the suggestion, made the change!